docs-src/typescript/how-to-handle-workflow-logic-requirements-in-typescript.md

```diff
@@ -0,0 +1,47 @@
+---
+id: how-to-handle-workflow-logic-requirements-in-typescript
+title: How to handle Workflow logic requirements in TypeScript
+sidebar_label: Workflow logic requirements
+description: Handle Signal
+tags:
+  - developer-guide
+  - sdk
+  - typescript
+---
+
+In the Temporal TypeScript SDK, Workflows run in a deterministic sandboxed environment.
+The code is bundled on Worker creation using Webpack, and can import any package as long as it does not reference Node.js or DOM APIs.
+
+:::note
+
+If you **must** use a library that references a Node.js or DOM API and you are certain that those APIs are not used at runtime, add that module to the [`ignoreModules`](https://typescript.temporal.io/api/interfaces/worker.BundleOptions#ignoremodules) list.
+
+:::
+
+Because the Workflow sandbox can run only deterministic code, side effects and access to external state must be done through Activities as Activity outputs are recorded in the Event History, and can read deterministically by the Workflow.
+
+This limitation also means that Workflow code cannot directly import the [Activity Definition](/concepts/what-is-an-activity-definition).
+[Activity Types](/concepts/what-is-an-activity-definition#activity-type) _can_ be imported, so they can be invoked in a type-safe manner.
```

Comment:
```suggestion
[Activity Types](/concepts/what-is-an-activity-definition#activity-type) *can* be imported, so they can be invoked in a type-safe manner.
```
Not sure why, but the underscores rendered literally; I replaced 'em with asterisks.